perlPackages.GlibObjectIntrospection: needs cairo to build test libs

livnev opened the pull request:
##### Motivation for this change

Without `cairo`, the build output has:

```
Trying to build test libraries... Package cairo was not found in the pkg-config search path.
Perhaps you should add the directory containing `cairo.pc'
to the PKG_CONFIG_PATH environment variable
No package 'cairo' found
 at Makefile.PL line 187.
not OK
Error: *** can not find package cairo
*** check that it is properly installed and available in PKG_CONFIG_PATH
 at Makefile.PL line 187.
Continuing without test library support...
```

and tests are skipped. Adding `cairo` to `buildInputs` makes the tests run, and fail! At least we know that this package is broken, with some issue related to library paths...

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

livnev committed:
perlPackages.GlibObjectIntrospection: needs cairo to build test libs

doronbehar commented on pkgs/top-level/perl-packages.nix:
```diff
@@ -7212,6 +7212,7 @@ let
       url = mirror://cpan/authors/id/X/XA/XAOC/Glib-Object-Introspection-0.047.tar.gz;
       sha256 = "1sk32pcv4ik3sniqbjrz8nlw7f2a76l87wx1rwdjypv6h9pjnrnk";
     };
+    buildInputs = [ pkgs.cairo ];
```
Why not add it to `checkInputs`?

livnev replied:
I changed it to `checkInputs`, fixed.

volth commented:
Is including `nextcloud: 18.0.0 -> 18.0.1` to this PR intentional?

livnev commented:
> Is including `nextcloud: 18.0.0 -> 18.0.1` to this PR intentional?

Nope, not at all. That somehow appeared after I rebased on latest master. I rebased on latest master again and now it's gone, so I think we're good. Sorry about that.

aanderse commented:
@GrahamcOfBorg build perldevelPackages.GlibObjectIntrospection perlPackages.GlibObjectIntrospection

aanderse commented:
@livnev looks like some issues still:
```
Result: FAIL
Failed 18/18 test programs. 0/0 subtests failed.
```

livnev commented:
@aanderse I am aware that the tests fail, my PR is adding a dependency which is necessary for the tests to run. I am suspecting that the package is broken but this was previously undetected since the tests were being skipped, due to the lack of `cairo`. Unfortunately I don't know how to fix it.

doronbehar commented:
I have a suspicion that the tests fail due to the sandbox or something in yours / hydra's environment. @livnev could you try and play with this with `nix-build --options sandbox false -A perlPackages.GlibObjectIntrospection`? If that fails as well, I'd contact upstream in the meantime and set `doCheck = false;`, along with the `checkInputs` added.

If the tests fail only with the sandbox, perhaps @aanderse we can disable the tests only if sandbox is enabled? Something like:

```
doCheck = !stdenv.isDarwin;
```

livnev committed:
perlPackages.GlibObjectIntrospection: mark as broken

livnev commented:
@doronbehar disabling sandboxing didn't fix it. For the record, all of the test failures look like this:

```
t/00-basic-types.t ............ WARNING **: Failed to load shared library '/nix/store/iaqncam6zxha0d1jgz5bnfhg3kbv5izj-perl5.30.1-Glib-Object-Introspection-0.048/lib/libregress.so' referenced by the typelib: /nix/store/iaqncam6zxha0d1jgz5bnfhg3kbv5izj-perl5.30.1-Glib-Object-Introspection-0.048/lib/libregress.so: cannot open shared object file: No such file or directory at /tmp/nix-build-perl5.30.1-Glib-Object-Introspection-0.048.drv-0/Glib-Object-Introspection-0.048/blib/lib/Glib/Object/Introspection.pm line 110.
Could not find GType for package Regress::TestObj at ./t/inc/setup.pl line 33.
Compilation failed in require at t/00-basic-types.t line 3.
BEGIN failed--compilation aborted at t/00-basic-types.t line 3.
```

I've now added `doCheck = false;` to this package.

doronbehar commented:
According to `nix-locate` (great tool for debugging these kind of issues BTW, see https://github.com/bennofs/nix-index ):

```
$ nix-locate libregress.so
gnome3.gjs.out                                  121,984 x /nix/store/4855j0vvm39rkcqdxdk8rnklj1gs00w2-gjs-1.56.2/lib/gjs/libregress.so
```

livnev commented:
> According to `nix-locate` (great tool for debugging these kind of issues BTW, see https://github.com/bennofs/nix-index ):
> 
> ```
> $ nix-locate libregress.so
> gnome3.gjs.out                                  121,984 x /nix/store/4855j0vvm39rkcqdxdk8rnklj1gs00w2-gjs-1.56.2/lib/gjs/libregress.so
> ```

Thanks for the hint. However, adding `pkgs.gnome3.gjs` to `propagatedBuildInputs` doesn't seem to fix the tests. Did you have something else in mind?

doronbehar commented:
I'd try to debug this inside a `nix-shell` while inspecting `$GI_TYPELIB_PATH` with this new dependency and perhaps try to run a perl command that will import this library somehow.

doronbehar commented:
See https://nixos.wiki/wiki/Nixpkgs/Create_and_debug_packages#Using_nix-shell_for_package_development

livnev commented:
@doronbehar unfortunately since I'm not a perl person I think I would struggle to fix this. I stumbled upon this when trying to get another package to build which depends on this one. This means I may have to abandon the PR, or it can be merged so that at least the breakage is visible.

volth commented:
In the current state of
```diff
+    checkInputs = [ pkgs.cairo ];
+    doCheck = false;
```
the PR has no sense.

Build is not broken on `master`, the check is just skipped due to missing `cairo`.

livnev commented:
> In the current state of
> 
> ```diff
> +    checkInputs = [ pkgs.cairo ];
> +    doCheck = false;
> ```
> 
> the PR has no sense.
> 
> Build is not broken on `master`, the check is just skipped due to missing `cairo`.

I was mainly trying to add visibility to the fact that the package is broken. Would you advise removing `doCheck = false;` in that case?

doronbehar commented:
I think this PR makes sense. @livnev I'd have also add a comment to the derivation with a link to this PR and a quote of the error: 

```
t/00-basic-types.t ............ WARNING **: Failed to load shared library '/nix/store/iaqncam6zxha0d1jgz5bnfhg3kbv5izj-perl5.30.1-Glib-Object-Introspection-0.048/lib/libregress.so' referenced by the typelib: /nix/store/iaqncam6zxha0d1jgz5bnfhg3kbv5izj-perl5.30.1-Glib-Object-Introspection-0.048/lib/libregress.so: cannot open shared object file: No such file or directory at /tmp/nix-build-perl5.30.1-Glib-Object-Introspection-0.048.drv-0/Glib-Object-Introspection-0.048/blib/lib/Glib/Object/Introspection.pm line 110.
Could not find GType for package Regress::TestObj at ./t/inc/setup.pl line 33.
Compilation failed in require at t/00-basic-types.t line 3.
BEGIN failed--compilation aborted at t/00-basic-types.t line 3.
```

While writing down as well that adding `gnome.gjs` doesn't fix this issue. Perhaps use a `# TODO` comment which will be `grep`able as well.

livnev commented:
@doronbehar sorry for the delay, done.

doronbehar commented:
Wait, now that I see https://github.com/NixOS/nixpkgs/issues/68116 , I can see you wrote there:

> I tried making a package that depends on GlibObjectIntrospection, and got the following error at runtime:
> ...

Meaning that the failure of the tests _correctly_ indicates the package is broken. I'd remove `doCheck = false;` and add to `meta`:

```
    broken = true;
```

And put that nice comment you added before `doCheck`, there.

livnev commented:
> Wait, now that I see #68116 , I can see you wrote there:
> 
> > I tried making a package that depends on GlibObjectIntrospection, and got the following error at runtime:
> > ...
> 
> Meaning that the failure of the tests _correctly_ indicates the package is broken. I'd remove `doCheck = false;` and add to `meta`:
> 
> ```
>     broken = true;
> ```
> 
> And put that nice comment you added before `doCheck`, there.

Agreed, I was trying to say that the tests failing was a good thing because the package actually appears to be broken. I've now reverted the `doCheck = false`, marked it as broken, and moved the comment. Thanks for your guidance on this.
